readme.md
# mup-node plugin

Deploy your Node.js apps with Meteor Up!

This plugin is under development and is missing some features.

## Getting started

First, install mup and mup-node with:

```bash
npm i -g mup mup-node
```

Second, create a config with
``bash
mup init
``

Open the config, and make the following adjustments:

For each server:
- host - Usually is the IP Address of the server
- server authentication - You can use a password or set pem to the path to a private key. If neither are set, it uses ssh-agent

In the `app` section:

- name: A unique name, with no spaces
- path: Path to the app, relative to the config.
- type: Set to `node` to let mup know that this plugin will manage the app

Add a `plugins` array with `mup-node`:

```js
module.exports = {
  // ... rest of config
  plugins: [ 'mup-node' ]
};
```

Third, setup the server. Mup will install everything needed to run the app. Run:

```bash
mup setup
```

Fourth, deploy the app. Run

```bash
mup deploy
```

Mup will upload your app, build a docker container, and run it.

## Bundling

`mup-node` copies the app's files to the server. The `.git` and `node_module` directories are ignored.Additional directories can be ignored by creating a `.mupignore` file in your app's root folder. `.mupignore` uses the same syntax as `.gitignore`.

## Building Image

The base image is `node:<node version>`. As long as the package.json doesn't change, the `node_modules` is cached between deploys.

You can add instructions to the Dockerfile in the config at `app.docker.buildInstructions`. For more details, look at the example in the next section.

If your package.json has a `mup:postinstall` script, it will be run after `npm install`.

## Options

`mup-node` is configured with the `app` object in your config. The available options are:

```js
module.exports = {
  app: {
    name: 'name-of-app',
    path: '../path/to/app',
    type: 'node',
    nodeVersion: '8.8.1',
    servers: {
      one: {},
      two: {}
    },
    env: {
      MONGO_URL: 'mongodb://localhost:8000',
      // Port the app is available on, defaults to 80
      PORT: 5000
    },
    // Amount of time to allow the app to start. If the app isn't running within this
    // much time, mup rolls back to the previous version. Set to -1 to disable.
    deployCheckWaitTime: 60,
    // NPM script to run when starting the app
    startScript: 'start:production'
    docker: {
      args: ['--network=net'],
      networks: ['net2', 'net3'],
      // Dockerfile instructions to run before adding the app and running `npm install`
      buildInstructions: [
        // Copy some of the app's files needed for the post-install script
        'COPY ./scripts ./scripts',
        'RUN apt-get update && apt-get install libcairo2-dev libjpeg-dev libpango1.0-dev libgif-dev build-essential g++ -y',
      ],
      // Port expose from the container. This does not affect the port the app is accessed on. Default is 3000
      imagePort: 3000
    }
  },
  plugins: [ 'mup-node' ]
};
```


changelog.md
## 0.8.2 April 20, 2022

- Update tar

## 0.8.1 July 7, 2021

- Allow overriding env for specific servers, similar to how mup does it for Meteor apps

## 0.8.0 October 23, 2020

- Add `mup node debug` command. This lets you connect your local Node dev tools, for example those in Chrome, to your Node app running on the server.

## 0.7.1 May 6, 2020

- Support load balancing with Mup 1.5

## 0.7.0 March 23, 2020

- Adds support for the `privateDockerRegistry` option
- Adds `mup node destroy` command to remove app from the servers

## 0.6.0 July 5, 2019

- `.mupignore` file can be used to exclude files and folders from being bundled
- Add `app.docker.imagePort` option, which works the same as in Meteor Up for meteor apps
- Remove validation warning for empty environment variables
- Fix plugin modifying config when app type was not 'node'

## 0.5.0 August 23, 2018

- **Breaking Change** Environment variables from config are set when building the docker image. Due to this, dev dependencies will no longer be installed by default.
- Copy package-lock.json into the docker container before installing dependencies
- Update instructions to include adding `mup-node` to the array of plugins

## 0.4.0 April 2, 2018

- Allow disabling Verifying Deployment
- Show warning when package.json is missing a start script
- Fix start task succeeding despite errors starting the container

## 0.3.0 - March 9, 2018

- Run the npm script `mup:postinstall` when building the image, if it exists

## 0.2.1 - March 7, 2018

- Fix setting default value for `docker.buildInstructions`

## 0.2.0 - March 7, 2018

- Add options to customize image
- Show defaults in config when `mup validate --show` is run


index.js
var handlers = require('./command-handlers');
var validator = require('./validate.js');

function onlyNodeEnabled(commandName) {
  return function (api) {
    var appConfig = api.getConfig().app;

    if (appConfig && appConfig.type === 'node') {
      return api.runCommand(commandName);
    }
  }
}

function setDefault(obj, path, value) {
  var position = obj;

  path.reduce((position, property, index) => {
    if (index === path.length - 1) {
      if (typeof position[property] === 'undefined') {
        position[property] = value;
      }
    } else {
      position[property] = position[property] || {};

      return position[property];
    }
  }, obj);

  return obj;
}

module.exports = {
  name: 'node',
  description: 'Deploy node.js apps',
  commands: {
    setup: {
      description: 'Setup node.js',
      handler: handlers.setup
    },
    deploy: {
      description: 'Deploy app',
      handler: handlers.deploy
    },
    reconfig: {
      description: 'Reconfig app',
      handler: handlers.reconfig
    },
    start: {
      description: 'Start app',
      handler: handlers.start
    },
    stop: {
      description: 'Stop app',
      handler: handlers.stop
    },
    logs: {
      description: 'View app\'s logs',
      handler: handlers.logs
    },
    destroy: {
      description: 'Remove app from server',
      handler: handlers.destroy,
      builder(subYargs) {
        return subYargs.option('force', {
          description: 'forces app to be removed',
          boolean: true
        });
      }
    },
    // hidden commands
    build: {
      description: false,
      handler: handlers.build
    },
    push: {
      description: false,
      handler: handlers.push
    },
    debug: {
      name: 'debug [server]',
      description: 'Debug the meteor app',
      builder(yargs) {
        yargs.strict(false);
      },
      handler: handlers.debugApp
    }
  },
  hooks: {
    'post.default.setup': onlyNodeEnabled('node.setup'),
    'post.default.reconfig': onlyNodeEnabled('node.reconfig'),
    'post.default.deploy': onlyNodeEnabled('node.deploy'),
    'post.default.logs': onlyNodeEnabled('node.logs'),
    'post.default.start': onlyNodeEnabled('node.start'),
    'post.default.stop': onlyNodeEnabled('node.stop')
  },
  prepareConfig(config) {
    var appConfig = config.app;

    if (!appConfig || appConfig.type !== 'node') {
      return config;
    }


    setDefault(appConfig, ['env', 'NODE_ENV'], 'production');
    setDefault(appConfig, ['env', 'PORT'], 80);

    setDefault(appConfig, ['docker', 'buildInstructions'], []);

    setDefault(appConfig, ['startScript'], 'start');
    setDefault(appConfig, ['nodeVersion'], 'latest');
    setDefault(appConfig, ['deployCheckWaitTime'], 60);

    return config;
  },
  validate: {
    app(config, utils) {
      if (config.app && config.app.type === 'node') {
        return validator(config, utils);
      }
      return [];
    }
  }
}
